developerportal/apps/staticbuild/wagtail_hooks.py

```diff
@@ -27,41 +32,153 @@ class ArticleAdmin(ModelAdmin):
+#   it, takes down the flag and starts a static build
+#
+# * The static build runs within a simple distributed lock, so if a build it
+#   still running when the periodic task runs and finds a new build has been
+#   requestd, we don't allow a concurrent build, but we do preserve/re-instate
+#   the flag ready for the next attempt by the periodic task. This means we only
+#   ever have one build in progress, but we can immediately follow a build with
+#   a new one if there's still an unsatisfied request for a build.
+
+EXPECTED_BUILD_AND_SYNC_JOB_FUNC_NAME = (
+    "developerportal.apps.staticbuild.wagtail_hooks._static_build_async"
+)
+
+SENTINEL_KEY_TIMEOUT = (
+    60 * 15
+)  # 15 mins - longer than a build-and-sync should be taking
+SENTINEL_KEY_NAME = "devportal-fresh-build-and-sync-needed"
+SENTINEL_LOCK_NAME = "deveportal-sentinel-lock"
+BUILD_LOCK_NAME = "deveportal-build-and-sync-lock"
+
+# Evaluate this now to avoid an awkward edge case during the former concurrent
+# builds where it (somehow) an os.path.join()ed value involving
+# settings.BUILD_DIR gets re-set with the joined value...
+BUILD_ROOT_DIR = settings.BUILD_DIR
+
+
 def _generate_build_path():
     """Generate a unique build path to avoid concurrent builds clashing"""
-    return os.path.join(settings.BUILD_DIR, tz_now().isoformat())
+    return os.path.join(BUILD_ROOT_DIR, tz_now().isoformat())
 
 
-@app.task
-def _static_build_async(force=False, pipeline=settings.STATIC_BUILD_PIPELINE):
-    """Calls each command in the static build pipeline in turn."""
-    log_prefix = "Static build task"
-    build_dir = None
+def _is_build_and_sync_job(job_details):
+    return job_details["name"] == EXPECTED_BUILD_AND_SYNC_JOB_FUNC_NAME
 
-    if settings.DEBUG is False or force:
-        build_dir = _generate_build_path()
-        logger.info(f"{log_prefix} Created temporary build dir {build_dir}")
 
-    for name, command in pipeline:
-        if settings.DEBUG and not force:
-            logger.info(f"{log_prefix} (wagtail-bakery) command '{name}' skipped.")
-        else:
-            logger.info(f"{log_prefix} (wagtail-bakery) command '{name}' started.")
-            # We're passing in a specific output dir so that we don't risk
-            # concurrent builds clashing
-            call_command(command, build_dir=build_dir)
-            logger.info(f"{log_prefix} (wagtail-bakery) command '{name}' finished.")
+def _set_build_needed_sentinel(oid):
+    """Idempotent flag that a static build should be attempted when next checked
+    (which is within STATIC_BUILD_JOB_ATTEMPT_FREQUENCY seconds)"""
+    with redis_lock(SENTINEL_LOCK_NAME, oid=oid):
+        cache.set(SENTINEL_KEY_NAME, True, SENTINEL_KEY_TIMEOUT)
+
+
+def _get_build_needed_sentinel(oid):
+    with redis_lock(SENTINEL_LOCK_NAME, oid=oid):
+        sentinel_val = cache.get(SENTINEL_KEY_NAME)
+        cache.delete(SENTINEL_KEY_NAME)
+    return sentinel_val
```

Comment:
@stevejalim Apologies again for being so slow to look at this! I like your overall approach! I've been skimming the details, and I had a question about the use here of `SENTINEL_LOCK_NAME`. It seems that lock is never checked anywhere to see whether or not it has been acquired? So it seems like the `SENTINEL_LOCK_NAME` lock is not really doing anything, or that it's not necessary, since it would never block the `cache.set(SENTINEL_KEY_NAME, ...` or `cache.get(SENTINEL_KEY_NAME)`. But I'm probably missing something!

Reply:
Heya @escattone - thanks for the laser eyes!  

You're right - while in the main use of redis_lock we check for whether we actually get the lock, here (and in `_set_build_needed_sentinel`) I'm not. (It was a last-minute bit of clean-up refactoring and I clearly didn't think it through.)

The goal in `_get_build_needed_sentinel` was to try to make `cache.get()` and `cache.delete()` atomic in , given there's no `cache.pop()` operation available. (Looking back, it should probably be called `_pop_build_needed_sentinel` for greaster clarity, too).

Given that this is a situation where we really do want to store the sentinel, rather than only store if there's no existing lock, I guess we could  either:

a) just remove use of redis_lock and take the (pretty minor?) risk of another process setting the cache between `.get()` and `.delete()` (I say pretty minor cos the static site gets rebuild hourly anyway...)

b) change this to be a blocking/waiting lock, but that might deadlock (until the `redis_lock` times out)

So, I think opening a small PR (given this is now merged) to remove the call to `redis_lock` in this getter/setter pair (and the related `SENTINEL_LOCK_NAME`) is the way to go. Does that make sense to you? 

Thanks again for the catch!

Reply:
@stevejalim Oh, I see, that totally makes sense that you'd want the `get` and `delete` calls within `_get_build_needed_sentinel` to be atomic. I think the [`MULTI`/`EXEC` transaction wrappers](https://redis.io/topics/transactions) that Redis provides may be exactly what you'd want here, but I don't think those commands can be made via the Django `cache` object. I think you'd have to use a dedicated Redis client like `redis-py`, which of course then you'd have to configure. However, you'd then have the full power of Redis available, including the `EVAL` command, which allows you to run entire scripts atomically (see https://redis.io/commands/eval#atomicity-of-scripts). I've used the `EVAL` command to handle locking in one of my projects(see https://github.com/escattone/walnut/blob/master/walnut/core.py if you'd like to see examples).

Reply:
@stevejalim If setting up a dedicated Redis client seems like more work than you think it's worth, your idea of removing the call to `redis_lock` in the getter/setter pair seems good to me as well. In other words, it's so tempting to be a purist, but it just may not be worth it! :)

Reply:
Yeap, we're def on the same page @escattone! Would be nice to do it 'right', but maybe it's not 100% needed... FWIW, i originally was using django-redis's `cache.lock()` which maps to `redis-py`'s own `Lock`, but thought it better to standardise on one lock implementation (and then fumbled the refactor :/)

FWIW here's the django-redis lock: https://niwinz.github.io/django-redis/latest/#_locks which might actually be a reasonable drop-in, now that the project is using `django-redis` anyway. `.lock` is basically a pass-through to `redis.py`

No need to reply - i know you're busy. I've made a new issue to come back v soon (P2, moving soon to P1) and address the point you've raised. Thanks for the support!

Reply:
@stevejalim Wow, didn't know about django-redis' `cache.lock()` and `redis-py`'s `Lock`! Really nice! Thank you!